man/man1/clive.1
.\" Process this file with
.\" groff -man -Tascii clive.1
.\"
.Dd April 10, 2002
.Dt CLIVE 1
.Sh NAME
.Nm clive
.Nd A console LiveJournal client
.Sh SYNOPSIS
.Nm clive
.Op Fl u Ar username
.Op Fl w Ar password
.Op Fl j Ar journal
.Op Fl s Ar subject
.Op Fl r Ar security
.Op Fl f Ar mood
.Op Fl i Ar userpic
.Op Fl m Ar music
.Op Fl l Ar location
.Op Fl t Ar taglist
.br
.Op Fl p
.Op Fl d
.Op Fl Fl backdate Ar YYYYMMDDhhmm
.Op Fl Fl replace= Ns Ar itemid
.Op Fl Fl charset Ar encoding
.Op Fl Fl plainpass | Fl Fl hashpass
.Op Fl Fl softreturn
.Pp
.Nm clive
.Op Fl Fl loginonly | Fl Fl nologin
.Nm clive
.Op Fl Fl lfgrp | Fl Fl laccess | Fl Fl lpic | Fl Fl lfriends | Fl Fl levents Ns Op = Ns Ar num
.Nm clive
.Op Fl Fl addfriend Ar name | Fl Fl delfriend Ar name
.Sh DESCRIPTION
.Nm clive
is a console client for the LiveJournal system.  It doesn't offer much
in the way of features.  It's strength lies in the ability to pipe the
text for an entry into it on stdin, thus allowing it to be used
filter-style. It an also lauch your preferred editor to type entries.
It allows for users to login and post/read/replace events, and that's
it.
.Sh OPTIONS
.Bl -tag -width Ds -compact
.It Fl u Ar username
.It Fl Fl user Ns = Ns Ar username
Specify the username
.It Fl w Ar password
.It Fl Fl password Ns = Ns Ar password
Specify the password
.It Fl j Ar journal
.It Fl Fl usejournal Ns = Ns Ar journal
Specify the journal to use, if different from the username.
You can also use the associated ID number obtainable with 
.Fl Fl laccess .
.It Fl s Ar subject
.It Fl Fl subject Ns = Ns Ar subject
The subject of the post.  This will disable interactive entry of the
subject.
.It Fl r Ar level
.It Fl Fl security Ns = Ns Ar level
Clive defaults to public security, unless specified otherwise.
This option is how you do that from the command line.
See the
.Sx SECURITY
section below for a description of that spec.
.It Fl f Ar mood 
.It Fl Fl mood Ns = Ns Ar mood
LiveJournal allows you to display a mood along with your post
Use this option to set it.
Use quotes if you want to have whitespace in your mood.
.It Fl m Ar music
.It Fl Fl music Ns = Ns Ar music
Describe the music you are listening to while writing this entry.
Use quotes if you want to have whitespace in your music title.
.It Fl l Ar location
.It Fl Fl location Ns = Ns Ar location
Describe where you are writing this entry (e.g. the name of your town).
.It Fl t Ar tag Ns Op Ns , Ns Ar tag Ns , Ns Ar ...
.It Fl Fl taglist Ar tag Ns Op Ns , Ns Ar tag Ns , Ns Ar ...
Add tags to categorise this entry.  Multiple tags must be separated by
commas.
.It Fl i Ar keyword
.It Fl Fl userpic Ar keyword
Many users have more than one user picture, you can specify which one
you would like to use here.  You can either use the picture keyword,
or the ID as specified by 
.Fl Fl lpic .
.It Fl p
.It Fl Fl preformat
This will tell LiveJournal that we are sending preformatted text.
This is useful if you have a lot of HTML in it.
.It Fl d
.It Fl Fl nocomments
This will tell LiveJournal to disable comments on this post.
.It Fl v
.It Fl Fl version
Display Clive version details and exit.
.It Fl Fl help
Display a short usage statement.
.It Fl Fl backdate Ar YYYYMMDDhhmm
Back date the item to year
.Ar YYYY ,
month
.Ar MM ,
day
.Ar DD ,
hour
.Ar hh ,
minute
.Ar mm .
This will insert it into your journal at that date, and it will prevent
the post from showing on other users' friends pages.
.It Fl Fl replace Ns = Ns Ar itemid
Replace an existing entry. The itemid can be obtained by running with
.Fl Fl levents
first. Note that not just the entire text, but all settings of the post
will be overridden, so you should use
.Fl Fl backdate
to reuse the old date. To delete an existing entry, just leave the event
text empty.
.It Fl Fl charset Ns = Ns Ar encoding
Specify the local character set (e.g.
.Ql utf-8 ,
.Ql iso8859-2 ,
.Ql koi8-r ) .
By default, the characterset is automatically detected (from environment
variables and system settings). Use this option to override the detected
value. This option can be set in the configuration file.
.It Fl Fl plainpass
Clive, by default, uses a challenge-response system to verify your
password.  If you use this option, it will send your password as
plaintext.
.Em Use of this option is not recommended.
.It Fl Fl hashpass
Clive, by default, uses a challenge-response system to verify your
password. If you use this option, it will send an MD5 hash of your
password to the server. Which is faster, but not very secure.
.Em  Use of this option is not recommended,
.It Fl Fl softreturn
This will collapse multiple lines to a single long line. A new paragraph
can still be started with an empty line in the text. In general this
will generate pretier output on the website. This option can be set in
the configuration file.
.It Fl Fl lfrgp
This will display a list of your friends groups with their associated ID
numbers.
.It Fl Fl laccess
This will display a list of the journals you have access to and their
associated ID numbers. If you join or leave a community, these numbers
will change.
.It Fl Fl lpic
This will dislpay a list of your userpics and their associated IDs.
If you make any changes to your user pictures, the IDs displayed may change.
.It Fl Fl lfriends
This will display a list of your friends and their birthdays (if set).
.It Fl Fl addfriend Ar name
Add a new friend. It is not an error to add an existing friend.
.It Fl Fl delfriend Ar name
Delete a friend. It is not an error to delete an already non-existing
friend.
.It Fl Fl levents Ns Op = Ns Ar num
This will display a list of the num most recent journal entries. If no
argument is given, it will only show the ID, timestamp and subject of
the most recent entry.
.It Fl Fl loginonly
.It Fl Fl nologin
These options control the behavior of clive with respect to logging into
the LJ server. Logging in allows you to retrieve any user-specific
information such as friend groups. If you wish to just login, use
.Fl Fl loginonly .
If you don't want to login at all use
.Fl Fl nologin .
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa ~/.cliverc
Your personal configuration file, you can set most options in here so
that you don't have to pass them in on the command line.
.El
.Sh SECURITY
LiveJournal has a variety of security settings: public, private, and
custom.  Clive supports all of these through the
.Op Fl r
or
.Op Fl Fl security
option.  Posts default to public security unless set on the commandline.
Here are the different security levels and how to specify them. 
.Bl -tag -width Ds
.It Ql private
Pass either 
.Ql 1
or 
.Ql private
as the argument, and you will be the only person that can view your post.
.It Ql friends
Pass either 
.Ql 2
or
.Ql friends
as the argument, and your post will only be viewable by the LiveJournal
users you list as friends.
.It Ql custom
LiveJournal allows you to specify up to 30 different friend groups,
and you can set your post readable by different groups. The way to
construct a security spec for a custom setting is to start with a
.Ql \&:
and then put in the name
.Bk
(e.g.
.Ql :closefriends ) .
.Ek
If you want to use multiple filters, continue separating with colons.
To include whitespace, use quotes.
.Bk
(e.g.
.Ql :schoolfriends:2:"Default View" )
.Ek
.El
.Sh HEADERS
It's annoying when you're editing a posting and realize another subject,
or mood description might be more appropriate. It is now possible to set
these in the posting itself by adding so called headers. If your posting
starts with lines containing special keywords followed by a colon and
value, these will be treated as special meta-information instead of the
body of your posting. For example currently available headers are:
.Pp
.Bd -literal -offset indent
Subject: An example posting
Mood: happy
Music: EAV - Ding Dong
Taglist: example,first post,test
.Ed
.Pp
These headers should be separated from the content of your posting by an
empty line.
.Sh CONFIGURATION
We understand that it can be annoying to pass in half a dozen command
line options every time you want to post to LiveJournal, so we allow
you to have a configuration file that specifies all of these.
You can specify key value pairs in your .cliverc for any of the command
line options that take values. Each pair should go on a line of it's
own. Any whitespace before and after the key or value is ignored. Lines
that begin with
.Ql #
are also ignored.  Each pair should look like this: 
.Dl key = value
.Sh ENVIRONMENT
.Bl -tag -width Ds -compact
.It Ev VISUAL
.It Ev EDITOR
The name for the external editor to use. If
.Ev VISUAL
is set, this will be preferred. If neither variable is set, a very lame
built-in editor will be used  This can be overriden in the configuration
file.
.El
.Sh SEE ALSO
.Pa http://ljclive.sourceforge.net/
.Sh AUTHORS
.Bl -item -compact
.It
Original author:
.An -nosplit
.An "Samuel Tesla"
.It
Current maintainer:
.An "Johan van Selst"
.It
Contact email:
.Aq clive-maintainer@gletsjer.net
.El


ChangeLog
Changes since 0.4.9:
- Add options -l (location) and -t (taglist; comma-separated)
- More consistent option handling and clarify manpage

Changes since 0.4.8:
- Fix memory allocation bug
- Convert (some) server information to local character set
  Note: this requires passing the --charset option
- Fix levents option by setting explicit protocol version

Changes since 0.4.7:
- Add server_port configuration option

Changes since 0.4.6:
- Add User-Agent to all HTTP requests, as required by LiveJournal

Changes since 0.4.5:
- Fix numerical security values
- Let connections be terminated immediately after response
- Improve server error message printing

Changes since 0.4.4:
- Fixed a bug in previous patch.
  Note UTF-8 only works for posting journal entries, not in other requests.

Changes since 0.4.3:
- Clean up and finalise the iconv() patch
  Use --charset to specify an input characterset (default: ISO-8859-1)

Changes since 0.4.2:
- Fix bug in prop_opt_preformatted that broke the -p option

Changes since 0.4.1:
- Change HTTP POST header to work with LiveJournal.com changes

Changes since 0.4.0:
- Add MD5 functions again for people without libmd

Changes since 0.3-alpha-p10:
- Improved check for missing functions
- New -v (--version) option
- New version numbering sceme: next release is 0.4.0

Changes since 0.3-alpha-p9:
- Removed compatibility functions (MD5, getopt_long)

Changes since 0.3-alpha-p8:
- Added BSD-style license - and install example cliverc in $pkgdatadir

Changes since 0.3-alpha-p7:
- Added --lcomments support (not in LJ yet)
- Added challenge/response authentication and made this default method
  Will automatically fallback to hash if server won't send a challenge
- Added --hashpass to use the old authentication

Changes since 0.3-alpha-p6:
- Added our own getopt_long() function, it's not as common as expected
- Added --enable-debug and --enable-DEBUG to ./configure script
- Add BUGS file with bug-reporting guidelines
- Several code cleanups
- Think I finally fixed the crasher

Changes since 0.3-alpha-p5:
- Worked around a crasher in Debian's asprintf() function
- Made a lot more function arguments constant
- Wrapped the source in GNU automake setup for easier distribution

Changes since 0.3-alpha-p4:
- Keep temporary data when posting fails (when using $EDITOR)
- Collapse multiple lines to a single line when --softreturn set
- Writing better documentation, error messages, source indenting

Changes since 0.3-alpha-p3:
- Fixed header detection (don't skip half the text)
- Fixed setting your mood (was broken a while ago)
- Fixed setting mood (with proper ID) from mood header

Changes since 0.3-alpha-p2:
- Use libc getopt() and getopt_long()
  If you don't have it: too bad - Try linking with -liberty then
- Interpret Subject: Music: and Mood: headers from the posting

Changes since 0.3-alpha-p1:
- Fixed custom post security using friendgroups
  It only works with names now, not with group IDs

Changes since 0.3-alpha:
- Decode the output of --levents
- Fixed a crasher in friend group listing
- Added --addfriend and --delfriend (only with default colours)

Changes since 0.2-beta-p1:
- Added --lfriends to list your friends (and their birthdays)
- Added --levents[=num] to list your recent journal entries
- Added --replace[=noteid] to replace/delete a journal entry
- Prompt for username and password if not supplied on cmdline/config



src/clive.c
/*
 * Copyright (C) 2001 - 2003, Samuel Tesla <stesla@pobox.com>
 * Copyright (C) 2003 - 2004, Johan van Selst <johans@stack.nl>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "system.h"
#include "net.h"
#include "util.h"
#include "lj.h"
#include "edit.h"
#include "configure.h"

#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>

int 
main(int argc, char **argv)
{
	lj_config	config;
	lj_server	server;
	lj_user		user;
	lj_logininfo	logininfo;
	lj_friendinfo	friendinfo;
	lj_eventinfo	eventinfo;
	lj_commentinfo	commentinfo;
	lj_event	post;
	lj_event_response post_response;

	if (lj_init_config(argc, argv, &config))
	{
		lj_error("Error initializing clive!\n");
	}
	lj_debug(2, "host: [%s]\n", config.server_host);
	lj_debug(2, "port: [%d]\n", config.server_port);
	lj_debug(2, "path: [%s]\n", config.server_path);

#ifdef		HAVE_ICONV
	utfenc = iconv_open("UTF-8", config.charset);
	utfdec = iconv_open(config.charset, "UTF-8");
	if ((iconv_t)-1 == utfenc || (iconv_t)-1 == utfdec)
		lj_error("Cannot open character conversion\n");
	lj_debug(2, "local charset: [%s]\n", config.charset);
#endif		/* HAVE_ICONV */


	/*******************
        ********************
        ** CLIENT VERSION **
        ********************
        *******************/

	memset(&logininfo, '\0', sizeof(logininfo));
	logininfo.clientversion = strdup("POSIX-Clive/" VERSION);
	if (config.getmoods && config.mood)
	{
		logininfo.mood = config.mood;
	}
	logininfo.userpics = NULL;
	if (config.getpickws)
		logininfo.loginflags |= getpickws;
	/* the serverinfo struct */
	server.hostname = config.server_host;
	server.port = config.server_port;
	server.path = config.server_path;

	/* the user struct */
	memset(&user, '\0', sizeof(user));
	user.username = config.username;
	user.password = config.password;
	if (config.plainpass)
		user.auth_method = pwd_plain;
	else if (config.hashpass)
		user.auth_method = pwd_hash;
	else
		user.auth_method = pwd_challenge;

	if (config.nologin)
	{
		if (config.loginonly)
		{
			lj_error("Make up your mind!  "
				"--nologin and --loginonly don't mix\n");
		}
	}
	else
	{
		printf("Logging on to %s, please wait\n", server.hostname);
		if (lj_login(&server, &user, &logininfo))
		{
			/* if this returns a non - zero value
			 * we know something broke
			 */
			if (logininfo.errmsg)
				lj_error("Errors logging in:\n\t%s\n",
					logininfo.errmsg);
			else
				lj_error("Errors logging in, try again later\n");
		}
		else
		{
			printf("Welcome to livejournal, %s\n", logininfo.name);
			if ((logininfo.message != NULL) &&
				strcmp(logininfo.message, ""))
			{
				fprintf(stderr, "A message from livejournal: %s\n",
					logininfo.message);
				if (isatty(fileno(stdin)))
				{
					puts("Press Enter to continue");
					getchar();
				}
			}
		}

		/* Handle non-posting events first */
		if (config.loginonly)
		{
			exit(0);
		}
		else if (config.lfrgp)
		{
			int	i;

			puts("ID\tGroup Name");
			for (i = 0; i < user.frgrpcount; i++)
			{
				printf("%2d\t%s\n", i, user.frgrps[i]);
			}
			exit(0);
		}
		else if (config.laccess)
		{
			int             i;

			puts("ID\tJournal Name");
			for (i = 0; i < user.accesscount; i++)
			{
				printf("%2d\t%s\n", i, user.access[i]);
			}
			exit(0);
		}
		else if (config.lpic)
		{
			int             i;

			puts("ID\tPicture Keyword");
			for (i = 0; i < user.userpiccount; i++)
			{
				printf("%2d\t%s\n", i,
				       user.userpic[i] ? user.userpic[i] : "(default)");
			}
			exit(0);
		}
		else if (config.lfriends)
		{
			int             i;

			memset(&friendinfo, 0, sizeof(friendinfo));

			lj_getfriends(&server, &user, &friendinfo);
			printf("%2d Friends\n", friendinfo.count);
			for (i = 0; i < friendinfo.count; i++)
			{
				printf("%3d %-10s %-14s  %s\n", i,
				       friendinfo.friends[i].birthday
					       ? friendinfo.friends[i].birthday
					       : "",
				       friendinfo.friends[i].user,
				       friendinfo.friends[i].name);
			}
			printf("%2d Friends of\n", friendinfo.focount);
			for (i = 0; i < friendinfo.focount; i++)
			{
				printf("%3d %-10s %-14s  %s\n", i,
				    friendinfo.friendsof[i].birthday
					    ? friendinfo.friendsof[i].birthday
					    : "",
				       friendinfo.friendsof[i].user,
				       friendinfo.friendsof[i].name);
			}
			exit(0);
		}
		else if (config.levents)
		{
			int             i;

			memset(&eventinfo, 0, sizeof(eventinfo));
			eventinfo.eventflags = config.levents;
			eventinfo.usejournal = config.usejournal;

			lj_getevents(&server, &user, &eventinfo);
			printf("%2d Events\n", eventinfo.count);
			for (i = 0; i < eventinfo.count; i++)
			{
				char 	*event = eventinfo.events[i].event;
				if (strlen(event) > 50)
					sprintf(event + 47, "...");
				printf("%3d %s  %-50.50s\n",
					eventinfo.events[i].itemid,
					eventinfo.events[i].time,
					eventinfo.events[i].event);
			}
			exit(0);
		}
		else if (config.lcomments)
		{
			int             i;

			memset(&commentinfo, 0, sizeof(commentinfo));
			commentinfo.journalid = config.lcomments;
			commentinfo.usejournal = config.usejournal;

			lj_getcomments(&server, &user, &commentinfo);
			printf("%2d Events\n", commentinfo.count);
			for (i = 0; i < commentinfo.count; i++)
			{
				char 	*comment = commentinfo.comments[i].comment;
				if (strlen(comment) > 50)
					sprintf(comment + 47, "...");
				printf("%3d %s  %s %-50.50s\n",
					commentinfo.comments[i].itemid,
					commentinfo.comments[i].time,
					commentinfo.comments[i].poster,
					comment);
			}
			exit(0);
		}
		else if (config.addfriend || config.delfriend)
		{
			memset(&friendinfo, 0, sizeof(friendinfo));
			friendinfo.friendflags =
				config.addfriend ? ADD_FRIEND : DELETE_FRIEND;
			friendinfo.count = 1;
			friendinfo.friends = malloc(sizeof(lj_friendninfo));
			memset(friendinfo.friends, 0, sizeof(friendinfo));
			friendinfo.friends[0].user = config.editfriend;

			lj_editfriends(&server, &user, &friendinfo);
			printf("%s %d friend%s\n",
				config.addfriend ? "Added" : "Deleted",
				friendinfo.count,
				friendinfo.count == 1 ? "" : "s");
			exit(0);
		}
	}


	/*
	 * Here we make sure that they have access to the journal they want
	 * to post to.
	 */
	post.usejournal = config.usejournal;

	if (user.accesscount && post.usejournal)
	{
		int             i = 0;

		if ((i = atoi(post.usejournal)))
			post.usejournal = user.access[i];

		for (i = 0; i < user.accesscount; i++)
			if (!strcmp(post.usejournal, user.access[i]))
				break;
		if (i >= user.accesscount)
			lj_error("You don't have access to post in %s!\n",
				post.usejournal);
	}

	post.nocomments = config.nocomments;
	post.preformatted = config.preformatted;
	post.music = config.music;
	post.location = config.location;
	post.taglist = config.taglist;
	if (config.backdated)
	{
		post.backdated = config.backdated;
		lj_strcopy(post.dateback, config.dateback);
	}
	else
	{
		post.backdated = 0;
		post.dateback[0] = '\0';
	}

	if (config.userpic)
	{
		int             i = -1;

		i = atoi(config.userpic);
		if (i >= 0)
		{
			char           *buff;

			buff = lj_malloc(strlen(config.userpic) + 1);
			sprintf(buff, "%d", i);
			if (!strcmp(buff, config.userpic))
			{
				/* we have a number ! */
				post.userpic = user.userpic[i];
			}
			else
			{
				post.userpic = config.userpic;
			}
			lj_free(buff);
		}
		else
		{
			post.userpic = config.userpic;
		}
	}
	else
	{
		post.userpic = NULL;
	}

	post.mood = NULL;
	post.moodid = 0;
	if (config.getmoods)
	{
		/* we have feelings ! */
		if (logininfo.moodid)
		{
			post.moodid = logininfo.moodid;
		}
		else if (logininfo.mood)
		{
			post.mood = strdup(logininfo.mood);
		}
	}
	if (config.customsecurity)
	{
		char		*prevp, *nextp;
		int		i;
		u_int32_t	mask = 0;

		/* skip the leading ':', then jump to following ':'s
		 * replacing them by '\0' to get seperate strings
		 */
		for (prevp = config.customsecurity + 1; prevp; prevp = nextp)
		{
			if ((nextp = strchr(prevp, ':')))
				*nextp++ = '\0';

			lj_debug(2, "Extracted [%s] from friendgroup spec\n",
				prevp);
			for (i = 0; i < user.frgrpcount; i++)
			{
				if (!strcmp(prevp, user.frgrps[i]))
				{
					lj_debug(2, "Anding [%d] to mask\n",
						i + 1);
					mask |= 1 << (i + 1);
					break;
				}
			}
			if (i == user.frgrpcount)
				lj_error("Invalid friendgroup. . . exiting\n");
		}
		post.security = strdup("usemask");
		asprintf(&post.allowmask, "%d", mask);
	}
	else
	{
		switch (config.security)
		{
		default:
			post.security = strdup("public");
			post.allowmask = strdup("");
			break;
		case 1:
			post.security = strdup("private");
			post.allowmask = strdup("");
			break;
		case 2:
			post.security = strdup("usemask");
			post.allowmask = strdup("1");
			break;
		}
	}
	post.replace = config.replace;

	/*
	 * we get these after everything else...that way we can be sure that
	 * any lj_error's come before the editor is launched.
	 */
	post.subject = get_subject(config.subject);
	if (post.subject && (strlen(post.subject) > LJ_SUBJECT_MAX))
	{
		lj_error("Subject exceeds allowed length. . . Exiting\n");
	}
	post.event = get_event(config.editor);

	/*
	 * do text postprocessing for meta information
	 */
	if (post.event)
	{
		char *start, *colon, *space, *end;

		start = post.event;
		while (*start && (end = strchr(start, '\n')))
		{
			if (!(colon = strchr(start, ':')))
				break;
			if ((start >= colon) || (colon >= end))
				break;
			if ((space = strchr(start, ' ')) && (colon > space))
				break;
			*end = '\0';
			*colon++ = '\0';
			while (*colon && isspace(*colon))
				colon++;
			if (!strcasecmp("Subject", start))
			{
				if (post.subject)
					free(post.subject);
				post.subject = strdup(colon);
			}
			else if (!strcasecmp("Music", start))
			{
				if (post.music)
					free(post.music);
				post.music = strdup(colon);
			}
			else if (!strcasecmp("Mood", start))
			{
				if (post.mood)
					free(post.mood);
				post.mood = strdup(colon);
			}
			else if (!strcasecmp("Location", start))
			{
				if (post.location)
					free(post.location);
				post.location = strdup(colon);
			}
			else if (!strcasecmp("Taglist", start))
			{
				if (post.taglist)
					free(post.taglist);
				post.taglist = strdup(colon);
			}
			else
			{
				printf("Ignoring unknown header %s\n", start);
			}
			start = end + 1;
		}
		if (config.softreturn)
			joinstring(start);
		post.event = start;
	}

	/* next we find our mood ID */
	if (post.mood && user.moodcount)
	{
		int i;
		post.moodid = 0;
		for (i = 0; i < user.moodcount; i++)
		{
			if (!strcmp(user.moods[i], post.mood))
				post.moodid = i;
		}
		lj_debug(2, "Found mood [%s] has moodid [%d]\n",
			post.mood, post.moodid);
	}


	if (!lj_postevent(&server, &user, &post, &post_response) &&
	    (post_response.message != NULL))
	{
		printf("Successfully posted item %s to journal %s \n",
		       post_response.message,
		       post.usejournal ? post.usejournal : user.username);
		remove_edit_file(1);
	}
	else if ((post_response.message != NULL)
		 && strcmp(post_response.message, ""))
	{
		printf("The following error was encountered "
			"while posting your message:\n"
			"\t%s\n",
			post_response.message);
		remove_edit_file(0);
	}
	else
	{
		puts("An unknown error was encountered "
			"while posting your message");
		remove_edit_file(0);
	}
	return 0;
}
